idx/memory/tag_query.go

```diff
@@ -134,34 +196,40 @@ func parseExpression(expr string) (expression, error) {
 }
 
 func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
-	q := TagQuery{from: from}
+	q := TagQuery{from: from, wg: &sync.WaitGroup{}}
 
 	if len(expressions) == 0 {
 		return q, errInvalidQuery
 	}
 
-	for _, expr := range expressions {
+	sort.Strings(expressions)
+	for i, expr := range expressions {
+		// skip duplicate expression
+		if i > 0 && expr == expressions[i-1] {
+			continue
+		}
+
 		e, err := parseExpression(expr)
 		if err != nil {
 			return q, err
 		}
 
 		// special case of empty value
 		if len(e.value) == 0 {
+			expression := kvRe{
+				key:        e.key,
+				value:      nil,
+				matchCache: &sync.Map{},
+				missCache:  &sync.Map{},
+			}
 			if e.operator == EQUAL || e.operator == MATCH {
-				q.notMatch = append(q.notMatch, kvRe{
-					key:   e.key,
-					value: nil,
-				})
+				q.notMatch = append(q.notMatch, expression)
 			} else {
-				q.match = append(q.match, kvRe{
-					key:   e.key,
-					value: nil,
-				})
+				q.match = append(q.match, expression)
 			}
 		} else {
-			// always anchor all regular expressions at the beginning
-			if (e.operator == MATCH || e.operator == NOT_MATCH) && e.value[0] != byte('^') {
+			// always anchor all regular expressions at the beginning if they do not start with ^
+			if (e.operator == MATCH || e.operator == NOT_MATCH || e.operator == MATCH_TAG) && e.value[0] != 94 {
```

Comment:
you can just use `'^'` instead of 94.
you can compile the below 2 programs with `go build -gcflags=-S` and see that the assembly is identical.
```
~/t/rune ❯❯❯ cat 1/main.go                                                                                                                                                                   ⏎
package main

import "fmt"

func main() {
	a := "^abc"
	fmt.Println(a[0] == 94)
}
~/t/rune ❯❯❯ cat 2/main.go
package main

import "fmt"

func main() {
	a := "^abc"
	fmt.Println(a[0] == '^')
}
```

Reply:
nice, i can also change the `switch / case` in `parseExpression()` then

Reply:
8186d8e